torch/optim/adadelta.py

```diff
@@ -118,6 +122,9 @@ def step(self, closure=None):
                 differentiable=differentiable,
             )
 
+        if not has_any_param_with_grad:
```

Comment:
The logic here can be combined with lines 108-110, right? In general, I believe this entire logic could go together under init_group instead of here.

However, before you implement any of those changes, a concern I have is that we're adding this same exact code to every optimizer. This is fragile because it is easy to forget to keep this code consistent across optimizers or when we add new ones. A better solution would be to use `register_optimizer_step_pre_hook` and register a pre-hook for all optimizers in optimizer.py. The hook has access to the optimizer, so you should be able to access the params with optimizer.param_groups. This WOULD trade off the performance as it adds an extra loop over the params.

So could we benchmark the following solutions across diff configs of the number of params (say 10, 100, 1000, 10000, etc):
1. Group the logic here into init_group
2. Registering a hook into that same logic

Reply:
@janeyx99 


Ran the simple benchmarking script below and got the following results


**With global pre-hook**

```
def _global_warn_step_no_param_with_grad_pre_hook(opt: Optimizer, args: Tuple[Any], kwargs: Dict[Any, Any]):
    any_param_has_grad = any([p.grad is not None for param_group in opt.param_groups for p in param_group['params']])
    if not any_param_has_grad:
        _warn_step_no_param_with_grad()

register_optimizer_step_pre_hook(_global_warn_step_no_param_with_grad_pre_hook)
```
Mapping from num_layers to time
{1: 0.005815193653106689, 10: 0.015650224685668946, 100: 0.09916568994522094, 1000: 1.0547807788848877}

**With current method (no extra loop over params)**

{1: 0.005666229724884033, 10: 0.014990105628967285, 100: 0.09507855415344238, 1000: 1.0070961976051331}

Does it suffice to say that the difference is significant enough that we would want to add any_param_has_grad as an optimizer attribute that can be accessed by the hook so we don't have to do the double loop?

<details><summary><b>Click for script</b></summary>
```
import torch
import torch.nn as nn
import time

torch.manual_seed(5)
num_layers = [1, 10, 100, 1000]
times = dict()

for num_layers in num_layers:
    sum_overall_step_time = 0
    for i in range(100):
        print(i)
        model = torch.nn.Sequential(*(nn.Linear(64, 64) for _ in range(num_layers)))
        inp = torch.randn(64, 64)
        opt = torch.optim.SGD(model.parameters(), lr=0.001)

        num_iters = 100
        overall_step_time = 0
        for i in range(num_iters):
            opt.zero_grad()
            out = model(inp)
            out.sum().backward()
            before_step = time.time()
            opt.step()
            after_step = time.time()
            step_time = after_step - before_step
            overall_step_time += step_time
        sum_overall_step_time += overall_step_time

    average_overall_step_time = sum_overall_step_time / 100
    times[num_layers] = average_overall_step_time
print(times)
``` 

</details>

Reply:
Yea

Reply:
I used a global post hook instead but I suppose there is still some duplicate code :/ what do you think?

Reply:
From offline discussion, perhaps the post hook isn't giving enough value in terms of code consistency here, since there is already a fair bit of duplicate code if we want to only loop over params once. Further, adding an attribute to optimizers so that the post hook can access it causes a lot of PT2 failures